### PR TITLE
Tmpfs mount point redirect

### PR DESCRIPTION
**Pull request by stevenhorsman**

### CODE REVIEW

File: aws/image/files/etc/systemd/system/kata-agent.service
Comment (yoheiueda) [inline aws/image/files/etc/systemd/system/kata-agent.service:15]: I think unmounitng the directories is necessary when the service is stopped.
```
ExecStopPost=umount /image
ExecStopPost=umount /kata-containers
```

Comment (stevenhorsman) [reply aws/image/files/etc/systemd/system/kata-agent.service:15]: Thanks, I'll add that in :)

Comment (stevenhorsman) [reply aws/image/files/etc/systemd/system/kata-agent.service:15]: Should be sorted now.

File: ibmcloud/image/files/etc/systemd/system/kata-agent.service
Comment (knrt10) [inline ibmcloud/image/files/etc/systemd/system/kata-agent.service:21]: no new line at the end of file.

**Review by knrt10 — CHANGES_REQUESTED** [review]
Small NIT

File: ibmcloud/image/files/etc/systemd/system/kata-agent.service
Comment (stevenhorsman) [reply ibmcloud/image/files/etc/systemd/system/kata-agent.service]: Thanks, should be fixed now :)

**Comment by bpradipt** [comment]
@stevenhorsman possible to add the changes for vsphere as well if it's not too much work?
It's perfectly fine if vsphere changes are handled in a separate PR.

**Comment by bpradipt** [comment]
> > @stevenhorsman possible to add the changes for vsphere as well if it's not too much work? It's perfectly fine if vsphere changes are handled in a separate PR.
> 
> That's cheeky Pradipta 😉 - I thought I was going crazy and the vsphere image code hadn't been merged yet, but it was 5mins ago. I'll add it to this PR in a few mins :)

Hope we won't need such kind of changes across the providers once @yoheiueda PR lands gets merged :-)

**Review by bpradipt — DISMISSED** [review]
/lgtm
Thanks @stevenhorsman

**Review by mkulke — COMMENTED** [review]
Since we use systemd mount files already, does it make sense to use them for the bind mounts too? Something like this:

```bash
cat /etc/systemd/system/run-image.mount
[Unit]
Description=Mount unit for /run/image
Before=kata-agent.service

[Mount]
What=/image
Where=/run/image
Type=none
Options=bind

[Install]
WantedBy=kata-agent.service
````

**Comment by stevenhorsman** [comment]
> Since we use systemd mount files already, does it make sense to use them for the bind mounts too? Something like this:
> 
> ```shell
> cat /etc/systemd/system/run-image.mount
> [Unit]
> Description=Mount unit for /run/image
> Before=kata-agent.service
> 
> [Mount]
> What=/image
> Where=/run/image
> Type=none
> Options=bind
> 
> [Install]
> WantedBy=kata-agent.service
> ```

Hey Magnus, I'm happy to give that a go and replace the kata-agent.service logic with this. I'm guessing that we wouldn't need to worry about unmounting with this approach rather than the kata-agent.service StartPre method as it's more decoupled from the agent service lifecycle? (Sorry if you can't tell I'm not a systemd expert!)

**Comment by mkulke** [comment]
> > Since we use systemd mount files already, does it make sense to use them for the bind mounts too? Something like this:
> > ```shell
> > cat /etc/systemd/system/run-image.mount
> > [Unit]
> > Description=Mount unit for /run/image
> > Before=kata-agent.service
> > 
> > [Mount]
> > What=/image
> > Where=/run/image
> > Type=none
> > Options=bind
> > 
> > [Install]
> > WantedBy=kata-agent.service
> > ```
> 
> Hey Magnus, I'm happy to give that a go and replace the kata-agent.service logic with this. I'm guessing that we wouldn't need to worry about unmounting with this approach rather than the kata-agent.service StartPre method as it's more decoupled from the agent service lifecycle? (Sorry if you can't tell I'm not a systemd expert!)

yup, I don't think we need to umount. One thing that can be surprising is the filename of systemd mounts, which is significant (replace `/` with `-` and hence some funky escaping).

**Comment by stevenhorsman** [comment]
@mkulke, Hey Magnus - sorry it took so long to get this in - one of those days. I've hopefully got it working and it seems to work on in the ibmcloud tests I did:
```
root@podvm-nginx-78440fe5:~# df -h /run/image/
Filesystem      Size  Used Avail Use% Mounted on
/dev/vdb1        98G  1.4G   92G   2% /run/image
root@podvm-nginx-78440fe5:~# df -h /run/kata-containers/
Filesystem      Size  Used Avail Use% Mounted on
/dev/vdb1        98G  1.4G   92G   2% /run/kata-containers
root@podvm-nginx-78440fe5:~# df -h /run/kata-containers/shared/containers/
Filesystem      Size  Used Avail Use% Mounted on
/dev/vdb1        98G  1.4G   92G   2% /run/kata-containers
```

Now that we don't need a tmpfs for shared containers I think we can remove the separate mount service for that as it's within the `/kata/containers` one, but I might be wrong about this?

**Review by bpradipt — APPROVED** [review]
/lgtm
Thanks @stevenhorsman

**Comment by mkulke** [comment]
> Now that we don't need a tmpfs for shared containers I think we can remove the separate mount service for that as it's within the `/kata/containers` one, but I might be wrong about this?

Sorry for the late review. I ran into a bit of trouble when attempting to run those changes on our internal CI (on azure). containers fail to start and I wasn't able to figure out why that happens yet. it might be an issue on our end, I'll look into it tomorrow.

```
2022/11/21 22:03:12 [adaptor/proxy] CreateContainer: calling PullImage for "docker.io/library/nginx:1.14.2" before CreateContainer (cid: "64424a9f498ab14222a015299f10b3f7387f8fde513ae59730620c4c09a1603d")
2022/11/21 22:03:13 [adaptor/proxy] CreateContainer: failed to call PullImage, probably because the image has already been pulled. ignored: rpc error: code = Internal desc = failed to mount "overlay" to "/run/kata-containers/nginx_1.14.2/rootfs", with error: EINVAL: Invalid argument
2022/11/21 22:03:13 [adaptor/proxy] CreateContainer fails: rpc error: code = Internal desc = failed to mount "/run/kata-containers/shared/containers/64424a9f498ab14222a015299f10b3f7387f8fde513ae59730620c4c09a1603d/rootfs" to "/run/kata-containers/64424a9f498ab14222a015299f10b3f7387f8fde513ae59730620c4c09a1603d/rootfs", with error: ENOENT: No such file or directory
```

**Review by knrt10 — APPROVED** [review]

**Comment by knrt10** [comment]
I haven’t tested it myself on Azure yet. So would suggest to wait until @mkulke review.

**Comment by bpradipt** [comment]
@mkulke were you able to test this on azure ? Can we merge it ?

**Comment by mkulke** [comment]
> @mkulke were you able to test this on azure ? Can we merge it ?

Please go ahead and merge, it's still failing on our CI. I don't spot anything problematic in the code and if it's tested on other platforms it should work like this on azure. we'll provide a fix if necessary.

**Review by mkulke — APPROVED** [review]

**Comment by bpradipt** [comment]
> > @mkulke were you able to test this on azure ? Can we merge it ?
> 
> Please go ahead and merge, it's still failing on our CI. I don't spot anything problematic in the code and if it's tested on other platforms it should work like this on azure. we'll provide a fix if necessary.

thanks @mkulke

**Pull request merged**
